Bluetooth: Controller: Fix BIS payload dereferencing and Enable PAST in HCI samples

cvinayak opened the pull request:
- Fix incorrect use of BIS indices to dereference the payload array, instead correctly use synchronised stream indices
- Fix compile error due to missing lll_df_types include
- Align Kconfig values used with bt-ll-sw-split Controller in hci_uart_3wire same as used in hci_uart sample
- Enabled Periodic Advertising Sync Transfer feature in the Bluetooth Low Energy HCI Controller samples
- Refactor the sample so that multiple advertising setup can be reused in other sample if required. Bluetooth enable can be in the sample that reuses the advertising setup
- Updates to project configuration extras to build with Zephyr Controller. Configure for 2 broadcast audio sources in 2 groups, 2 unicast audio source in 1 group and, 2 audio sinks for 1 group for both broadcast and unicast usecases.

cvinayak committed:
samples: Bluetooth: hci_uart hci_ipc configuration update

Updates to project configuration extras to build with
Zephyr Controller.

Configure for 2 broadcast audio sources in 2 groups,
2 unicast audio source in 1 group and, 2 audio sinks
for 1 group for both broadcast and unicast usecases.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ISO Sync Receiver BIS payload dereferencing

Fix incorrect use of BIS indices to dereference the payload
array, instead correctly use synchronised stream indices.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: iso_broadcast: Restrict max stream sync

Restrict possible maximum stream synchronization to
implementation defined 2 stream, as structure declarations
in the implementation is limited to 2.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: iso_broadcast: Fix buf alloc timeout

Fix buf alloc timeout to 20 millisecond as the sent callback
can have a latency that is upto twice the ISO interval.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -648,14 +655,14 @@ static void isr_rx(void *param)
 
 		/* Get reference to stream context */
-		stream_handle = lll->stream_handle[lll->stream_curr];
+		stream_handle = lll->stream_handle[stream_curr];
```
`stream_curr` is a generic and undocumented field, so it's hard to understand what it is. 

Is it an index starting from 0 to `num_bis`? 

How does it relate to the other (undocumented) fields?

- Is `num_bis` the number of BIS currently synced to, or the number of BIS in the BIG? 
- What is the relationsship between `num_bis` and `stream_count`?
- How does it related to `bis_curr`? 
- Somewhat related, is `bis_idx` related to the BIS indexes (starting from 0x01), or is another index for BIS unrelated to the BIS indexes? 

In the changes below you have a check `if ((stream_curr < BT_CTLR_SYNC_ISO_STREAM_MAX) &&` which indicitates the `stream_curr` can sometimes be larger than `BT_CTLR_SYNC_ISO_STREAM_MAX`, but that check is only there any not anywhere where you use it directly as an index to `stream_handle`.

cvinayak replied:
> Is it an index starting from 0 to num_bis?

`stream_curr` hold 0 indexed stream that is currently synchronized to. It ranges upto the number of streams required to be sync-ed. i.e. `0 <= stream_curr < stream_count`  where `0 < stream_count <= num_bis`.

> Is num_bis the number of BIS currently synced to, or the number of BIS in the BIG?

the number of BIS in the BIG

> What is the relationsship between num_bis and stream_count?

`0 < stream_count <= num_bis`

> How does it related to bis_curr?

`bis_curr` is the currently synchronized BIS amongst the total BIS from the Broadcaster.

> Somewhat related, is bis_idx related to the BIS indexes (starting from 0x01), or is another index for BIS unrelated to the BIS indexes?

yes, `bis_idx` is the BIS indexes (starting from 0x01) and local-variable corresponding to `bis_curr` in the LLL context structure.


> In the changes below you have a check if ((stream_curr < BT_CTLR_SYNC_ISO_STREAM_MAX) && which indicitates the stream_curr can sometimes be larger than BT_CTLR_SYNC_ISO_STREAM_MAX, but that check is only there any not anywhere where you use it directly as an index to stream_handle.

The check is added to handle compiler warning that the `payload` array could be accessed out-of-bound using `stream_curr` (because the compiler thinks `stream_count` could have greater value than `BT_CTLR_SYNC_ISO_STREAM_MAX`

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -553,6 +554,9 @@ static void isr_rx(void *param)
 		/* BIS index */
 		bis_idx = lll->bis_curr - 1U;
 
+		/* Current stream */
+		stream_curr = lll->stream_curr;
+
```
`stream_curr` is never used before it gets overridden in line 595

cvinayak replied:
There is a goto just below here, when the reception is done without a PDU being received (PDU missed for the current stream).

Thalley replied:
Ah, missed that. Damn `goto`s :D

cvinayak replied:
github diff can hide context...

Regarding use of `goto` needs some thought... hopefully I have used it for common exit paths and not complicated readability.

Nice read here: https://koblents.com/Ches/Links/Month-Mar-2013/20-Using-Goto-in-Linux-Kernel-Code/

Thalley replied:
Yeah, it was probably a mix of `goto`, Github diff and the size of the function (in lines of code) :)

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -816,14 +823,15 @@ static void isr_rx(void *param)
 		/* Next selected stream */
-		if (stream_curr < lll->stream_count) {
+		stream_curr = lll->stream_curr + 1U;
+		if ((stream_curr < BT_CTLR_SYNC_ISO_STREAM_MAX) &&
+		    (stream_curr < lll->stream_count)) {
```
Seems redundant to have both checks, as I assume that `stream_count` is never larger than `BT_CTLR_SYNC_ISO_STREAM_MAX`?

cvinayak replied:
See my previous reply, this is related to handling compiler warning.

Thalley replied:
Aha. Interesting. 

Could it be, in both functions if needed, replaced by an ASSERT? 

If `stream_curr` ever gets larger than `BT_CTLR_SYNC_ISO_STREAM_MAX`, or rather if `lll->stream_count` ever gets larger than `BT_CTLR_SYNC_ISO_STREAM_MAX` that's a significant issue that shouldn't silent be ignored like this check seemingly does.

cvinayak replied:
`stream_count` is assigned after parameter validation in `ll_big_sync_create()`, the check here with `BT_CTLR_SYNC_ISO_STREAM_MAX`  is already redundant.

Thalley replied:
> `stream_count` is assigned after parameter validation in `ll_big_sync_create()`, the check here with `BT_CTLR_SYNC_ISO_STREAM_MAX` is already redundant.

Do that mean you'll remove it, or what's the plan?

cvinayak replied:
This is the warning I get for bsim builds GCC 14.2.1 (not for target build using zephyr sdk), what do you suggest?

```
[121/198] Building C object zephyr/subsys/bluetooth/controller/ll_sw/nordic/CMakeFiles/subsys__bluetooth__controller__ll_sw__nordic.dir/lll/lll_sync_iso.c.obj
/home/jack/workspace/zephyrproject/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c: In function ‘isr_rx’:
/home/jack/workspace/zephyrproject/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:833:59: warning: array subscript 1 is above array bounds of ‘uint16_t[1]’ {aka ‘short unsigned int[1]’} [-Warray-bounds=]
  833 |                         stream_handle = lll->stream_handle[stream_curr];
      |                                         ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
In file included from /home/jack/workspace/zephyrproject/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:32:
/home/jack/workspace/zephyrproject/zephyr/subsys/bluetooth/controller/ll_sw/nordic/../../../controller/ll_sw/lll_sync_iso.h:75:18: note: while referencing ‘stream_handle’
   75 |         uint16_t stream_handle[BT_CTLR_SYNC_ISO_STREAM_MAX];
      |                  ^~~~~~~~~~~~~
/home/jack/workspace/zephyrproject/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:869:50: warning: array subscript 1 is above array bounds of ‘struct node_rx_pdu *[1][28]’ [-Warray-bounds=]
  869 |                                 if (!lll->payload[stream_curr][payload_index]) {
      |                                      ~~~~~~~~~~~~^~~~~~~~~~~~~
/home/jack/workspace/zephyrproject/zephyr/subsys/bluetooth/controller/ll_sw/nordic/../../../controller/ll_sw/lll_sync_iso.h:77:29: note: while referencing ‘payload’
   77 |         struct node_rx_pdu *payload[BT_CTLR_SYNC_ISO_STREAM_MAX]
      |                             ^~~~~~~
[196/198] Linking C executable zephyr/zephyr.elf
```

Thalley replied:
@cvinayak does you get that when `BT_CTLR_SYNC_ISO_STREAM_MAX >= 2`? 

The whole "next BIS" part of the function should arguably be guarded by `BT_CTLR_SYNC_ISO_STREAM_MAX >= 2` right? If `BT_CTLR_SYNC_ISO_STREAM_MAX == 1` then is there any need to check more than 1 BIS?

cvinayak replied:
> does you get that when BT_CTLR_SYNC_ISO_STREAM_MAX >= 2?

yes, the native sim target compiler for some reason does not know that `lll->stream_count` would not exceed, hence, it complains about out of bounds access.

> does "Next BIS" refer to the next BIS in the BIG, and not the next BIS that we've synced to?

"Next BIS" *finds* the next BIS that we've synced to, to schedule for radio reception at the right time of the subevent.

I have removed the additional check which is not complained by the target gcc, let me see if CI complains if any.

Thalley commented:
@cvinayak any updates for this? If it's planned for 4.0 we should probably get it ready soon :)

cvinayak commented:
> @cvinayak any updates for this? If it's planned for 4.0 we should probably get it ready soon :)

i will try to follow up tonight, but this is a bug and can go in later too with a bug report.

@larsgk is there any chance someone can get this verified.  Brief history, this is a fix related to BIS headsets being built with lower stream count support and still be able to synchronize to broadcasters with multiple streams. This was a discord discussion back then

cvinayak committed:
samples: Bluetooth: hci_uart_3wire: Align Kconfigs with hci_uart

Align Kconfig values used with bt-ll-sw-split Controller as
same used in hci_uart sample.

Relates to commit 2c6306d09954 ("Bluetooth: Controller:
BT_CTLR_ISO_TX_BUFFER_SIZE from BT_ISO_TX_MTU")'.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: hci_uart(_3wire)/hci_ipc: Enable PAST feature

Enabled Periodic Advertising Sync Transfer feature in the
Bluetooth Low Energy HCI Controller samples.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix compile error when BT_CTLR_PRIVACY disabled

Fix compile error when BT_CTLR_PRIVACY is disabled. Compile
error seen building hci_uart sample for nrf54l15dk.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix compile error missing lll_df_types include

Fix compile error due to missing lll_df_types include.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: broadcaster_multiple: Refactor for reuse

Refactor the sample so that multiple advertising setup can
be reused in other sample if required. Bluetooth enable can
be in the sample that reuses the advertising setup.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
